### PR TITLE
0.10 Timestamp Support

### PR DESCRIPTION
**Pull request by yingnanliu**
This is our take on v 0.10.x support (message timestamps).

It basically implemented what @dasch mentioned in https://github.com/zendesk/ruby-kafka/issues/270#issuecomment-242332398, plus `FetchRequest` also has `api_version` set to 2.

### CODE REVIEW

**Comment by dasch** [comment]
:heart:

**Comment by dasch** [comment]
I think this looks good. I'd like to park it here until I have a better sense of when ruby-kafka will start officially supporting the new Kafka 0.10 protocol, at which point I will probably abandon support for Kafka 0.9.

File: lib/kafka/protocol/message.rb [hunk 1]
```diff
@@ -71,14 +72,15 @@ def self.decode(decoder)
...
+        timestamp = message_decoder.int64
         key = message_decoder.bytes
         value = message_decoder.bytes
 
         # The codec id is encoded in the three least significant bits of the
         # attributes.
         codec_id = attributes & 0b111
 
-        new(key: key, value: value, codec_id: codec_id, offset: offset)
+        new(key: key, value: value, codec_id: codec_id, offset: offset, create_time: Time.at(timestamp/1000))
```
Comment (asoules) [inline lib/kafka/protocol/message.rb]: This should divide by a float in order to preserve milliseconds (ie. `1000.0`)

For example:
```
$ irb
$ timestamp = 1473867722801 # milliseconds
 => 1473867722801
$ Time.at(timestamp/1000).strftime('%S.%N')
 => "02.000000000"
$ Time.at(timestamp/1000.0).strftime('%S.%N')
 => "02.801000118"
```

Comment (yingnanliu) [inline lib/kafka/protocol/message.rb]: Thanks! It's fixed.

**Comment by thijsc** [comment]
It should be possible to support both 0.9 and 0.10 at the same time right?

**Comment by dasch** [comment]
@thijsc not easily, I'm afraid. Only in version 0.10 did Kafka introduce a way for clients to discover broker versions, and even then it's a bit arduous.

**Comment by thijsc** [comment]
Right. I do know other clients offer a configuration option for this. Is that an option you would consider?

**Comment by dasch** [comment]
Not at the moment - the other clients do this by having the user configure the protocol version manually, which I think is problematic. Once ruby-kafka moved to only support 0.10+ sometime in the future I will revisit this.

**Comment by niels-s** [comment]
@dasch any news on when this can be merged?

**Comment by dasch** [comment]
Once I'm in a position to thoroughly test ruby-kafka against 0.10 I'll create a stable branch for 0.9 support and move active development to 0.10. Hopefully within the next few months.

**Comment by hoanhp** [comment]
can anyone show me the proper config for that??

**Comment by dasch** [comment]
@hoanhp I'm not sure what you mean...

**Comment by hoanhp** [comment]
@dasch  sorry for my poor English.  I mean can you tell me how to custom kafka-broker's config to add timestamp into each message??

**Comment by dasch** [comment]
@hoanhp it's not currently supported by ruby-kafka – it will be in a future release targeting Kafka 0.10+.

**Comment by dasch** [comment]
@yingnanliu can you rebase this on the latest master?

I've switched master development to target Kafka 0.10 – compatibility with Kafka 0.9 will happen in the `v4.0-stable` branch, and only include bug fixes and security patches.

File: lib/kafka/protocol/produce_response.rb [hunk 2]
```diff
@@ -43,13 +45,16 @@ def self.decode(decoder)
               partition: decoder.int32,
               error_code: decoder.int16,
               offset: decoder.int64,
+              timestamp: decoder.int64,
```
Comment (dasch) [inline lib/kafka/protocol/produce_response.rb]: Maybe this should also be converted to a Time object?

**Comment by dasch** [comment]
@yingnanliu I hope it's okay if I take over this branch – I'd like to get this merged sooner rather than later.

**Pull request closed without merging**